lib/service/service.go

```diff
@@ -2339,52 +2358,40 @@ func (process *TeleportProcess) initSSH() error {
 				return trace.Wrap(err)
 			}
 			log.Infof("Service is starting in tunnel mode.")
+			waitFn = agentPool.Wait
+
+			process.OnExit("agentpool.shutdown", func(payload interface{}) {
+				agentPool.Stop()
+			})
 
 			// Broadcast that the node has started.
 			process.BroadcastEvent(Event{Name: NodeSSHReady, Payload: nil})
 		}
 
-		// Block and wait while the node is running.
-		s.Wait()
-		if conn.UseTunnel() {
-			agentPool.Wait()
-		}
-
-		log.Infof("Exited.")
-		return nil
-	})
+		// Cancel deferred cleanup, as we'll register an OnExit handler
+		ok = true
 
-	// Execute this when process is asked to exit.
-	process.OnExit("ssh.shutdown", func(payload interface{}) {
-		if payload == nil {
-			log.Infof("Shutting down immediately.")
-			if s != nil {
+		// Execute this when process is asked to exit.
+		process.OnExit("ssh.shutdown", func(payload interface{}) {
```

Comment:
It's tradition to do it like this, but there's nothing inherently special about `OnExit`. We could just block waiting for `process.WaitForEvent(TeleportExitEvent)` inline and do the cleanup afterwards right in the `ssh.node` callback, leaving some (or most?) of the defers to do the work. WDYT?

Reply:
That might be a bit simpler. Will give this a shot.